### PR TITLE
fixed undefined collections after `FOREACH`

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
`FOREACH` backups and restores variables and collections it sets for its `loopRoutine` so that everything is restored to the state it was before the `FOREACH`.

I did not check for `undefined` because I thought it would be fine to just restore it as undefined. But technically there is a slight difference between actually setting a variable to value `undefined` or just not setting it at all.

This PR only restores variables and collections if they were not `undefined` before the `FOREACH`. I did not test this one single bit so please do some testing.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3598/pr-test (or any other room on that server)

**Comment by bjalder26** [comment]
You already did my first PR idea 6 days before I had the idea.  Now you fixed this issue 6 hours before I discovered it.  I guess I'm getting closer. :)

**Comment by ArnoldSmith86** [comment]
I just reacted to #590. I had to look it up; I was convinced just now that you opened it. :smile:

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -619,9 +619,11 @@ export class Widget extends StateManaged {
           }
           await this.evaluateRoutine(a.loopRoutine, variables, collections, (depth || 0) + 1, true);
           for(const add in addVariables)
-            variables[add] = variableBackups[add];
+            if(variableBackups[add] !== undefined)
+              variables[add] = variableBackups[add];
           for(const add in addCollections)
-            collections[add] = collectionBackups[add];
+            if(collectionBackups[add] !== undefined)
+              collections[add] = collectionBackups[add];
```
Comment (rogerl50) [inline client/js/widgets/widget.js:626]: This leaves `DEFAULT` as it was inside the `FOREACH`, which isn't correct. It should be reset to `undefined`. (Same probably applies to the variables.)

**Review by rogerl50 — CHANGES_REQUESTED** [review]

File: client/js/widgets/widget.js (hunk 1 above)
Comment (ArnoldSmith86) [reply client/js/widgets/widget.js:626]: Huh. Thanks for catching.

**Review by rogerl50 — APPROVED** [review]
Approved through both test and code review.

**Pull request merged**